doc/index.rst

```diff
@@ -107,6 +107,15 @@ gives access to many of the documentation strings included in the code.
 Installation and download
 =========================
 
+Supported Python versions
+-------------------------
+
+The :mod:`uncertainties` package supports all versions of Python supported by
```

Comment:
It's true that **at this time** we've selected to support 3.8-3.12, largely because those are the version currently supported by python. However, I don't think the intention was to tie ourselves to always supporting the same version as Python. On the contrary I think we should support old versions until it becomes costly/annoying to do so for some reason or another. 

I think it would suffice to say 

> The :mod:`uncertainties` package supports Python 3.8 - 3.12. Supporting earlier Python versions will be considered if there is user demand. Versions of the :mod:`uncertainties` package up through 3.1.7 supported Python 2.7 through Python 3.12.

Reply:
Okay, I am fine with changing the text like that. Maybe we should confirm that we want to state 3.8+ here? Syntax-wise the code still supports 2.7+ but with #199 it would become difficult to package and install in 2.7, so I would like to at least officially drop 2.7, which is why I made this separate PR to clarify how we state Python version support (there is also some code clean up that can done after dropping 2.7 which I didn't start on here). I think the relevant reasons for bumping are:

* 3.0-3.4: I don't think 3.0-3.3 were widely used, so jumping to 3.4 seems okay to me after dropping 2.7.
* 3.5: aysnc/await syntax, probably not relevant
* 3.6: f-strings, which are nice for development
* 3.7: `from __future__ import annotations`, which converts type annotations to strings so syntax from later Python verisons can be used in type annotations -- nice for adding type annotations
* 3.8: assignment expressions (`:=`); can be useful but more something to use when available rather than something to break backwards compatibility for
* 3.9: Nothing that I know of
* 3.10: pattern matching (`match` statement) -- similar to `:=` where it is useful but not a big impact that requires breaking compatibility.
* 3.11-3.12: Nothing that I know of

So syntax-wise maybe f-strings and type annotations are reason enough to jump to 3.7 at least.

In terms of dependencies, there are only numpy and pytest currently, and there could be deprecations in the standard library modules (mainly `math` and `inspect` for uncertainties).

Overall though, no individual features are really compelling to me for dropping support for older versions, except maybe type annotations and 3.7. It is more just that the newer versions add some nice-to-haves and over time the user base of the older versions gets very small and makes it seem not worth avoiding using those newer features.

Reply:
I vote for testing with only Python 3.8 and later, and saying that we cannot claim that the library works on any system that we do not test.

I would vote for adding an f-string somewhere in the code, perhaps making https://github.com/lmfit/uncertainties/blob/1857e870e6626b8262e2d70f5cf91973fe5ac828/uncertainties/core.py#L1869 read

```python
        return f"{self.nominal_value}+/-{std_def_err}"
```

or (but, as with other conversation, maybe this comes later?):
 
```python
        return f"ufloat({self.nominal_value}, {std_def_err})"
```

That will deliberately break Python 3.5 and earlier.  

I would be okay with saying "We don't test or support Python 3.7, but it might still work for now... until it doesn't.".